catalog/app/containers/NavBar/Search.tsx

```diff
@@ -0,0 +1,362 @@
+      {...props}
+      startAdornment={
+        <M.InputAdornment position="start">
+          {iconized && !expanded ? (
+            <M.Icon className={classes.inputIcon} onClick={onHelpToggle}>
+              search
+            </M.Icon>
+          ) : (
+            <Lab.ToggleButton
+              className={classes.inputOptions}
+              size="small"
+              value="help"
+              selected={helpOpen}
+              onChange={onHelpToggle}
+              disabled={props.disabled}
+            >
+              <M.Icon fontSize="small">search</M.Icon>
+              <M.Icon fontSize="small">
+                {helpOpen ? 'arrow_drop_up' : 'arrow_drop_down'}
+              </M.Icon>
+            </Lab.ToggleButton>
+          )}
+        </M.InputAdornment>
+      }
+    />
+  )
+}
+
+function useSearchUrlState(bucket: string | undefined) {
+  const { paths, urls } = NamedRoutes.use()
+  const location = useLocation()
+  const match = useRouteMatch(paths.search)
+  const isInStack = BucketConfig.useIsInStack()
+  const settings = CatalogSettings.use()
+
+  const qs = match && parse(location.search, true)
+
+  const query = qs?.q ?? ''
+  const mode = qs?.mode ?? settings?.search?.mode
+  // if not in stack, search all buckets
+  const buckets = qs?.buckets ?? (bucket && isInStack(bucket) ? bucket : undefined)
+
+  const makeUrl = React.useCallback(
+    (q: string | null) => urls.search({ q, buckets, mode }),
+    [urls, buckets, mode],
+  )
+
+  const bucketList = React.useMemo(() => buckets?.split(',') ?? [], [buckets])
+
+  return { query, makeUrl, buckets: bucketList }
+}
+
+interface SearchState {
+  buckets: string[]
+  input: SearchInputProps
+  help: Pick<Parameters<typeof SearchHelp>[0], 'open' | 'onQuery'>
+  onClickAway: () => void
+}
+
+interface SearchProps extends SearchInputExternalProps {
+  bucket?: string
+  onFocus?: () => void
+  onBlur?: () => void
+}
+
+function useSearchState({ bucket, onFocus, onBlur, ...props }: SearchProps): SearchState {
+  const history = useHistory()
+
+  const { query, makeUrl, buckets } = useSearchUrlState(bucket)
+
+  const [value, change] = React.useState<string | null>(null)
+  const [expanded, setExpanded] = React.useState(false)
+  const [helpOpen, setHelpOpen] = React.useState(false)
+
+  const onChange = React.useCallback((evt: React.ChangeEvent<HTMLInputElement>) => {
+    change(evt.target.value)
+  }, [])
+
+  const handleExpand = React.useCallback(() => {
+    if (expanded) return
+    change(query)
+    setExpanded(true)
+    onFocus?.()
+  }, [expanded, query, onFocus])
+
+  const handleCollapse = React.useCallback(() => {
+    change(null)
+    setExpanded(false)
+    setHelpOpen(false)
+    onBlur?.()
+  }, [onBlur])
+
+  const handleHelpOpen = React.useCallback(() => setHelpOpen(true), [])
+
+  const handleHelpClose = React.useCallback(() => setHelpOpen(false), [])
+
+  const onQuery = React.useCallback(
+    (strPart: string) => change(`${value} ${strPart}`),
```

Comment:
Just a micro-optimizaton:
`onQuery` is triggered every `value` change, but Help actually can be independent of `value`
```tsx
const onQuery = useCallback((strPart) => setValue((v) => `$x{v} ${strPart}`), [])
```